XenAdmin/Dialogs/ControlDomainMemoryDialog.cs

```diff
@@ -0,0 +1,177 @@
+            // out of sync and maximum < minimum.
+            if (vm.memory_dynamic_max >= vm.memory_dynamic_min &&
+                vm.memory_static_max >= vm.memory_static_min)
+            {
+                double min = vm.memory_static_min;
+                double max = Math.Min(host.memory_available_calc, MAXIMUM_DOM0_MEMORY_GB * Util.BINARY_GIGA);
+                double value = vm.memory_dynamic_min;
+                // Avoid setting the range to exclude the current value: CA-40041
+                if (value > max)
+                    max = value;
+                if (value < min)
+                    min = value;
+                memorySpinner.Initialize(Messages.CONTROL_DOMAIN_MEMORY_LABEL, null, value, max);
+                memorySpinner.SetRange(min, max);
+            }
+            origMemory = memorySpinner.Value;
+            UpdateMaintenanceWarning();
+        }
+
+        private bool HasChanged()
+        {
+            return memorySpinner.Value != origMemory;
+        }
+
+        private bool SaveChanges()
+        {
+            if (!HasChanged())
+                return false;
+
+            var mem = memorySpinner.Value;
+
+            DialogResult dialogResult;
+            using (var dlg = new ThreeButtonDialog(
+                    new ThreeButtonDialog.Details(SystemIcons.Warning, Messages.CONFIRM_CHANGE_CONTROL_DOMAIN_MEMORY, Messages.XENCENTER),
+                    ThreeButtonDialog.ButtonYes,
+                    ThreeButtonDialog.ButtonNo))
+            {
+                dialogResult = dlg.ShowDialog(this);
+            }
+
+            if (DialogResult.Yes != dialogResult)
+                return false;
+
+            var actions = new List<AsyncAction>();
+            var action = new ChangeControlDomainMemoryAction(host, (long)mem, false);
+            actions.Add(action);
+
+            actions.Add(new RebootHostAction(host, AddHostToPoolCommand.NtolDialog));
+
+            var multipleAction = new MultipleAction(connection, 
+                string.Format(Messages.ACTION_CHANGE_CONTROL_DOMAIN_MEMORY, host.Name), 
+                string.Format(Messages.ACTION_CHANGE_CONTROL_DOMAIN_MEMORY, host.Name), 
+                Messages.COMPLETED, actions, true, false, true);
+
+            multipleAction.RunAsync();
+            return true;
+        }
+
+        private void Cleanup()
+        {
+            host.PropertyChanged -= Server_PropertyChanged;
+            if (hostMetrics != null)
+                hostMetrics.PropertyChanged -= Server_PropertyChanged;
+        }
+
+        private void Server_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            Program.Invoke(this, Populate);
+        }
+
+        private void OkButton_Click(object sender, EventArgs e)
+        {
+            if (SaveChanges())
+            {
+                DialogResult = DialogResult.OK;
+            }
+            Close();

```

Comment:
Here and in line 174 below: you shouldn't need to call Close, setting DialogResult (and both buttons already have a value set in the designer) should be ok to close it and return from ShowDialog.